crypto/bls12/bls12.go

```diff
@@ -155,231 +131,294 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
 	return agg.participants
 }
 
-// AddSignatures adds additional signatures to the aggregate.
-func (agg *AggregateSignature) AddSignatures(signatures map[hotstuff.ID]*Signature) {
-	g2 := bls12.NewG2()
-	for id, s := range signatures {
-		g2.Add(&agg.sig, &agg.sig, s.s)
-		agg.participants.Add(id)
-	}
+func firstParticipant(participants consensus.IDSet) hotstuff.ID {
+	id := hotstuff.ID(0)
+	participants.RangeWhile(func(i hotstuff.ID) bool {
+		id = i
+		return false
+	})
+	return id
 }
 
-// bls12Crypto is a Signer/Verifier implementation that uses bls12-381 aggregate signatures.
-type bls12Crypto struct {
+type bls12Base struct {
 	mods *consensus.Modules
-}
 
-// New returns a new bls12-381 signer and verifier.
-func New() consensus.CryptoImpl {
-	bc := &bls12Crypto{}
-	return bc
+	mut sync.RWMutex
+	// popCache caches the results of popVerify for each public key.
```

Comment:
Should mention here and probably other places that `pop` = `proof of possession`